core/trino-main/src/main/java/io/trino/sql/planner/SplitSourceFactory.java

```diff
@@ -173,11 +175,17 @@ private SplitSource createSplitSource(TableHandle table, Map<Symbol, ColumnHandl
                 dynamicFilter = dynamicFilterService.createDynamicFilter(session.getQueryId(), dynamicFilters, assignments);
             }
 
-            Constraint constraint = filterPredicate
-                    .map(predicate -> filterConjuncts(predicate, expression -> !DynamicFilters.isDynamicFilter(expression)))
-                    .map(predicate -> new LayoutConstraintEvaluator(plannerContext, session, assignments, predicate))
-                    .map(evaluator -> new Constraint(TupleDomain.all(), evaluator::isCandidate, evaluator.getArguments())) // we are interested only in functional predicate here, so we set the summary to ALL.
-                    .orElse(alwaysTrue());
+            Expression nonDynamicFilter = filterConjuncts(filterPredicate.orElse(TRUE), expression -> !DynamicFilters.isDynamicFilter(expression));
+            LayoutConstraintEvaluator evaluator = new LayoutConstraintEvaluator(plannerContext, session, assignments, nonDynamicFilter);
+
+            // we are interested only in functional predicate here, so we set the summary to ALL.
+            Constraint constraint = new Constraint(
+                    TupleDomain.all(),
+                    ConnectorExpressionTranslator.translateConjuncts(session, nonDynamicFilter).connectorExpression(),
+                    assignments.entrySet().stream()
+                            .collect(toImmutableMap(entry -> entry.getKey().name(), Entry::getValue)),
+                    evaluator::isCandidate,
```

Comment:
I think this is differing from the contract specified in `io.trino.spi.connector.Constraint#predicate`
```
    * A predicate that can be used to filter data. If present, it is equivalent to, or stricter than, {@link #getSummary()} and different from, and should be AND-ed with, {@link #getExpression()}
```
`io.trino.spi.connector.Constraint#getExpression`
```
an expression predicate which is different from, and should be AND-ed with, {@link #getSummary} or {@link #predicate} (if present).
```
It looks like what we're doing now is providing the same predicate through `evaluator::isCandidate` and ConnectorExpression.
It will help to change the javadocs and clarify that these can be the same thing

Reply:
The iceberg connector is using `io.trino.spi.connector.Constraint#predicate` today to filter splits based on partition values when there is a non-TupleDomain filter. I think it would be interesting to be able to use non-TupleDomain predicates to prune splits based on file min/max as well and not just partition values. e.g. https://github.com/trinodb/trino/issues/26038
But that requires translating the ConnectorExpression to iceberg expression.